File: 494. Target Sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        dp = {}
        def recurtion(idx, cur_sum):
            if idx == len(nums):
                return 1 if cur_sum == target else 0
            if (idx, cur_sum) in dp:
                return dp[(idx, cur_sum)]

            add = recurtion(idx + 1, cur_sum + nums[idx])
            subtract = recurtion(idx + 1, cur_sum - nums[idx])
            dp[(idx, cur_sum)] = add + subtract
            return add + subtract
        return recurtion(0, 0)

File: 494. Target Sum.cpp
class Solution {
public:
    map<pair<int,int>,int>dp;
    int solve(int n, int sum, vector<int>& nums)
    {
        if(!n) return sum == 0;

        if(dp.find({n,sum}) != dp.end()) return dp[{n, sum}];

        int add = solve(n - 1, sum - nums[n-1], nums);
        int subtract = solve(n - 1, sum + nums[n - 1], nums);
        return dp[{n, sum}] = add + subtract;
    }
    int findTargetSumWays(vector<int>& nums, int target) 
    {
        return solve(nums.size(), target, nums);
    }
};